### PR TITLE
U-mode: reset to initial state on unrecoverable error.

### PR DESCRIPTION
**Pull request by glg-rv**
The first patch allows to retrieve HypMap information at run time.

The second patch implement reset in umode. When the error is unrecoverable (Panic or UnhandledTrap), the hypervisor will:

1. restore the U-mode memory for the local CPU at initial state.
2. restore the U-mode architectural state for the local CPU at initial state.
3. run the newly reset umode until it initializes itself.

If we can't reset U-mode the hypervisor will fail an unwrap(). Because we won't be able to run U-mode again and just retrying would be just pointless.

### CODE REVIEW

File: src/hyp_map.rs [hunk 1]
```diff
@@ -209,6 +209,44 @@ impl PrivateRegion {
...
+            // Write to user mapping setting SUM in SSTATUS.
+            CSR.sstatus.modify(sstatus::sum.val(1));
+            // Safety:
+            // - this write is in a umode regions guaranteed to be mapped by HypMap in every page table.
+            // - the region starts at self.vaddr and is self.size byte long. `len` is <= `self.size`.
+            unsafe {
+                core::ptr::copy(data.as_ptr(), self.vaddr.bits() as *mut u8, len);
+            }
+            // Restore SUM.
+            CSR.sstatus.modify(sstatus::sum.val(0));
```
Comment (dgreid) [inline src/hyp_map.rs:233]: Is SUM guaranteed to start at 0 when running `restore`?

File: src/hyp_map.rs [hunk 2]
```diff
@@ -209,6 +209,44 @@ impl PrivateRegion {
...
+        // Copy data at the beginning if it's present.
+        if let Some(data) = self.data {
+            // In case data is bigger than region size, write up to region end only.
+            let len = core::cmp::min(self.size, data.len());
+            let data = &data[0..len];
+            // Copy original data to umode area.
+            // Write to user mapping setting SUM in SSTATUS.
+            CSR.sstatus.modify(sstatus::sum.val(1));
+            // Safety:
+            // - this write is in a umode regions guaranteed to be mapped by HypMap in every page table.
```
Comment (dgreid) [inline src/hyp_map.rs:227]: s/regions/region/

File: src/umode.rs [hunk 3]
```diff
@@ -255,33 +255,49 @@ impl UmodeTask {
...
+        let ret = task.run();
+        if let Err(e) = &ret {
+            match e {
+                Error::UnexpectedTrap | Error::Panic => {
+                    // Reset on unrecoverable u-mode error.
+                    // 1. restore memory to original state
+                    PerCpu::this_cpu().page_table().restore_umode();
+                    // 2. setup umode again.
+                    // Unwrap: if we can't restore umode the hypervisor is in an unrecoverable state.
+                    task.reset().unwrap();
```
Comment (dgreid) [inline src/umode.rs:295]: Can this unwrap be an expect with a good error message?

**Review by dgreid — COMMENTED** [review]
Just a few questions and nits, thank!

File: src/main.rs [hunk 4]
```diff
@@ -477,10 +462,20 @@ extern "C" fn kernel_init(hart_id: u64, fdt_addr: u64) {
...
     // Create the hypervisor mapping from the hardware memory map and the U-mode ELF.
-    let hyp_map = HypMap::new(mem_map, &umode_elf).expect("Cannot create Hypervisor map.");
+    HypMap::init(mem_map, &umode_elf).expect("Cannot create Hypervisor map.");
 
-    // The hypervisor mapping is complete. Can setup paging structures now.
-    setup_hyp_paging(hyp_map, &mut hyp_mem);
+    // Create per-cpu page tables.
+    let cpu_info = CpuInfo::get();
+    for i in 0..cpu_info.num_cpus() {
+        // Unwrap okay. HypMap has been initialised above.
```
[the remark this answers is not in the document]
Comment (glg-rv) [reply src/main.rs]: Whops. Sorry, artifact from simpler times. Fixed.

File: src/hyp_map.rs (hunk 1 above)
Comment (glg-rv) [reply src/hyp_map.rs:233]: It's up to what we mean by guarantee, I suppose. Sstatus is set to zero in salus, and we only use it here. Every time we modify it (when we enter guest or u-mode) we save it and restore it.

File: src/hyp_map.rs (hunk 2 above)
Comment (glg-rv) [reply src/hyp_map.rs]: Thanks, fixed this and the one a couple of lines below.

File: src/umode.rs (hunk 3 above)
Comment (glg-rv) [reply src/umode.rs]: Tried to make it a good message. Fixed.

File: src/hyp_map.rs [hunk 5]
```diff
@@ -209,6 +209,44 @@ impl PrivateRegion {
...
+            copied = len;
+        }
+        // Clear data from the end of copy to the end of mapped_area.
+        let len = mapped_size - copied;
+        let dest = self.vaddr.bits() + copied as u64;
+        // Write to user mapping setting SUM in SSTATUS.
+        CSR.sstatus.modify(sstatus::sum.val(1));
+        // Safety:
+        // - this write is in a umode region guaranteed to be mapped by HypMap in every page table.
+        // - writing to this region start at offset `copied` and goes untill the mapped size of the region.
```
Comment (atulkharerivos) [inline src/hyp_map.rs:243]: s/untill/until

**Review by atulkharerivos — APPROVED** [review]
LGTM barring minor nits

**Review by dgreid — APPROVED** [review]

File: src/umode.rs [hunk 6]
```diff
@@ -255,33 +255,49 @@ impl UmodeTask {
...
-        // return the error.
+        let ret = task.run();
+        if let Err(e) = &ret {
+            match e {
+                Error::UnexpectedTrap | Error::Panic => {
+                    // Reset on unrecoverable u-mode error.
+                    // 1. restore memory to original state
+                    PerCpu::this_cpu().page_table().restore_umode();
+                    // 2. setup umode again.
+                    // Unwrap: if we can't restore umode the hypervisor is in an unrecoverable state.
```
[the remark this answers is not in the document]
Comment (glg-rv) [reply src/umode.rs]: FIxed.

File: src/hyp_map.rs (hunk 5 above)
Comment (glg-rv) [reply src/hyp_map.rs]: Fixed.

**Pull request merged**